vowpalwabbit/io_buf.cc

```diff
@@ -158,3 +158,26 @@ void io_buf::close_file_or_socket(int f)
   close(f);
 #endif
 }
+
+//make all data that was writen available for reading
+void io_buf::flip_from_write_to_read()
```

Comment:
It looks like this is only used in tests? I'm not sure this operation actually makes any sense in real usage (at least according to the documentation in io_buf.h)

```
/* The i/o buffer can be conceptualized as an array below:
**  _______________________________________________________________________________________
** |__________|__________|__________|__________|__________|__________|__________|__________|   **
** space.begin           space.head             space.end                       space.endarray **
**
** space.begin     = the beginning of the loaded values in the buffer
** space.head      = the end of the last-read point in the buffer
** space.end       = the end of the loaded values from file
** space.endarray  = the end of the allocated space for the array
**
** The values are ordered so that:
** space.begin <= space.head <= space.end <= space.endarray
```

Reply:
Also it's not clear whether this operation is needed. If I remember I think I added it to hack around the way I was using io_buf in the test

Reply:
This method is only used by tests. I don't see the issue with that by itself.

There's no other way through the current API to flip an io_buf from "write mode" to "read mode".

Peter, it makes sense given that, in "write mode", `end() == begin()`. The documentation you mention only explain how io_buf works when reading.

Reply:
For example, say you write 10 bytes, you end up with the following set of values (assume data is the v_array storage:
```
end == begin == &data[0]
head == &data[10]
```

To read that we need the following set of pointers:

```
head == begin == &data[0]
end == &data[10]
```

This could potentially be used kernel_svm.

Reply:
If the documentation is only valid for read-mode, can you add another piece of documentation describing how write-mode works?

Reply:
the name of the function could be a bit misleading, feels like its a permissions switch were you are not allowed to write before calling this function.

Reply:
I don't think we should support flipping from write/read especially since it isn't needed right now either....

Reply:
I've removed this in the latest iteration